### ISSUE TITLE
Add `Event::AddressChange`.

### ISSUE THREAD

**Issue by dvc94ch** [quinn-rs/quinn#1095]
There is no way to detect remote address changes due to connection migration, but it can be polled using the `remote_address` method. It would be nice to have an event for it.

**Comment by Ralith** [comment]
That's definitely feasible. We'd need to decide whether to emit it on every attempted migration, or only after path validation succeeds. Can you share your use case to give the issue some context?

**Comment by dvc94ch** [comment]
For nat traversal you need to tell the peer what it's address is so that it can announce it's listening address correctly, so it is useful to be notified when it changes. The muxer trait in libp2p supports a `StreamMuxerEvent::AddressChange` event for this purpose. [0]

- [0] https://docs.rs/libp2p/0.36.0/libp2p/core/muxing/enum.StreamMuxerEvent.html

**Comment by Ralith** [comment]
Migration is initiated by receipt of a packet from a new source address. If the server is behind a NAT, wouldn't that packet be dropped?

**Comment by dvc94ch** [comment]
I guess the idea is that you have a connection open to a hole punching server. You might loose the connection to the server behind a firewall which drops incoming packets from an unknown source.

**Comment by dvc94ch** [comment]
the hole punching server might use it. but I guess it doesn't make a big difference. if you lose the connection you start again. and in the case of roaming you can detect the network change yourself.
